### PR TITLE
[Enhancement] Use shared file reader when read a segment

### LINKED ISSUE #10895: [Enhancement] Use shared FileReader when read a segment
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

In the current implementation, the process of reading a segment will repeatedly create multiple `FileReader`s for the same segment file, which will bring some unnecessary overhead (i.e., when creating S3FileReader, it will launch a GetHead request to get the size of S3 object).
We can only create `FileReader` once in `Segment::open`, other data structures related to reading this segment can use the same `FileReader`. Because `FileReader:: read_at` operation is thread-safe, so we can read the same `FileReader` safely.

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by platoneko**
# Proposed changes

Issue Number: close #10895 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/io/fs/file_reader.h [hunk 1]
```diff
@@ -38,7 +40,11 @@ class FileReader {
     virtual const Path& path() const = 0;
 
     virtual size_t size() const = 0;
+
+    virtual bool closed() const = 0;
 };
 
+using FileReaderPtr = std::shared_ptr<FileReader>;
```
Comment (yiguolei) [inline be/src/io/fs/file_reader.h:47]: if it is a shared ptr, FileReaderSPtr is better. If it is a unique ptr, FileReaderUPtr is better.

File: be/src/olap/rowset/beta_rowset_writer.h [hunk 2]
```diff
@@ -86,7 +86,7 @@ class BetaRowsetWriter : public RowsetWriter {
     std::unique_ptr<segment_v2::SegmentWriter> _segment_writer;
     mutable SpinLock _lock; // lock to protect _wblocks.
     // TODO(lingbin): it is better to wrapper in a Batch?
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_writer.h:88]: remove this comment.

File: be/src/io/fs/file_reader.h (hunk 1 above)
Comment (yiguolei) [inline be/src/io/fs/file_reader.h:47]: Have you test that there is no performance lost when change to use shared ptr?

File: be/src/olap/rowset/segment_v2/column_reader.h [hunk 3]
```diff
@@ -64,7 +64,7 @@ struct ColumnReaderOptions {
 };
 
 struct ColumnIteratorOptions {
-    io::FileReader* file_reader = nullptr;
+    io::FileReaderPtr file_reader = nullptr;
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_reader.h:67]: the pointer is called too many times, shared ptr may have performance issues

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 4]
```diff
@@ -349,7 +348,7 @@ Status SegmentIterator::_init_return_column_iterators() {
             ColumnIteratorOptions iter_opts;
             iter_opts.stats = _opts.stats;
             iter_opts.use_page_cache = _opts.use_page_cache;
-            iter_opts.file_reader = _file_reader.get();
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:352]: I think the previous code maybe want to improve the performance by using raw pointer.

File: be/src/io/fs/local_file_system.cpp [hunk 5]
```diff
@@ -56,28 +47,16 @@ Status LocalFileSystem::create_file(const Path& path, std::unique_ptr<FileWriter
     return Status::OK();
 }
 
-Status LocalFileSystem::open_file(const Path& path, std::unique_ptr<FileReader>* reader) {
+Status LocalFileSystem::open_file(const Path& path, FileReaderPtr* reader) {
     auto fs_path = absolute_path(path);
-    std::shared_ptr<OpenedFileHandle<int>> file_handle(new OpenedFileHandle<int>());
-    bool found = _file_cache->lookup(fs_path.native(), file_handle.get());
```
Comment (yiguolei) [inline be/src/io/fs/local_file_system.cpp:62]: file cache is useless?

File: be/src/io/fs/local_file_system.cpp [hunk 6]
```diff
@@ -56,28 +47,16 @@ Status LocalFileSystem::create_file(const Path& path, std::unique_ptr<FileWriter
...
-    }
     size_t fsize = 0;
     RETURN_IF_ERROR(file_size(fs_path, &fsize));
-    *reader = std::make_unique<LocalFileReader>(std::move(fs_path), fsize, std::move(file_handle));
+    int fd = -1;
+    RETRY_ON_EINTR(fd, open(fs_path.c_str(), O_RDONLY));
+    if (fd < 0) {
+        return Status::IOError("cannot open {}: {}", fs_path.native(), std::strerror(errno));
+    }
+    *reader = std::make_unique<LocalFileReader>(std::move(fs_path), fsize, fd);
```
Comment (yiguolei) [inline be/src/io/fs/local_file_system.cpp:59]: reader is shared ptr？ why call make unique here?

File: be/src/io/fs/local_file_reader.cpp [hunk 7]
```diff
@@ -39,16 +34,18 @@ LocalFileReader::~LocalFileReader() {
 }
 
 Status LocalFileReader::close() {
-    bool expected = false;
-    if (_closed.compare_exchange_strong(expected, true)) {
-        _file_handle.reset();
-        DorisMetrics::instance()->local_file_open_reading->increment(-1);
+    if (!closed()) {
```
Comment (yiguolei) [inline be/src/io/fs/local_file_reader.cpp:37]: Close method maybe called concurrently, the _fd maybe set to -1 in another thread. I think we'd better use lock here.

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 8]
```diff
@@ -17,7 +17,8 @@
 
 #include "olap/rowset/segment_v2/column_reader.h"
 
-#include "gutil/strings/substitute.h"                // for Substitute
+#include "gutil/strings/substitute.h" // for Substitute
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:20]: Substitue should not be used. Could remove this include.

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 9]
```diff
@@ -149,7 +149,7 @@ Status ColumnReader::read_page(const ColumnIteratorOptions& iter_opts, const Pag
                                BlockCompressionCodec* codec) const {
     iter_opts.sanity_check();
     PageReadOptions opts;
-    opts.file_reader = iter_opts.file_reader;
+    opts.file_reader = iter_opts.file_reader.get();
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:152]: opts.file_reader is raw pointer?

File: be/src/olap/rowset/segment_v2/page_io.h [hunk 10]
```diff
@@ -46,7 +46,7 @@ namespace segment_v2 {
 
 struct PageReadOptions {
     // block to read page
-    doris::io::FileReader* file_reader = nullptr;
+    io::FileReader* file_reader = nullptr;
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/page_io.h:49]: why not use shared pointer here?

File: be/src/io/fs/local_file_system.cpp (hunk 6 above)
Comment (platoneko) [reply be/src/io/fs/local_file_system.cpp:59]: I will fix it soon.

File: be/src/olap/rowset/segment_v2/page_io.h (hunk 10 above)
Comment (platoneko) [reply be/src/olap/rowset/segment_v2/page_io.h:49]: read_and_decompress_page is a synchronized method, this FileReader won't be released during reading page now.

File: be/src/olap/rowset/beta_rowset_writer.h (hunk 2 above)
Comment (platoneko) [reply be/src/olap/rowset/beta_rowset_writer.h:88]: done

**Review by yiguolei — APPROVED** [review]
lgtm

**Pull request merged**